modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp

```diff
@@ -382,32 +382,43 @@ void WindowContextBase::process_mouse_scroll(GdkEventScroll* event) {
     jdouble dx = 0;
     jdouble dy = 0;
 
-    // converting direction to change in pixels
-    switch (event->direction) {
-#if GTK_CHECK_VERSION(3, 4, 0)
-        case GDK_SCROLL_SMOOTH:
-            //FIXME 3.4 ???
-            break;
+    bool is_smooth = false;
+
+#ifdef GLASS_GTK3
+        if (gdk_window_get_events(gdk_window) & GDK_SMOOTH_SCROLL_MASK) {
+            is_smooth = true;
+            if (event->direction == GDK_SCROLL_SMOOTH) {
+                dx = event->delta_x;
+                dy = event->delta_y;
+            }
+        }
 #endif
-        case GDK_SCROLL_UP:
-            dy = 1;
-            break;
-        case GDK_SCROLL_DOWN:
+
+    if (!is_smooth) {
+        // converting direction to change in pixels
+        if (event->direction == GDK_SCROLL_UP) {
             dy = -1;
-            break;
-        case GDK_SCROLL_LEFT:
-            dx = 1;
-            break;
-        case GDK_SCROLL_RIGHT:
+        } else if (event->direction == GDK_SCROLL_DOWN) {
+            dy = 1;
+        } else if (event->direction == GDK_SCROLL_LEFT) {
             dx = -1;
-            break;
+        } else if (event->direction == GDK_SCROLL_RIGHT) {
+            dx = 1;
+        }
     }
+
     if (event->state & GDK_SHIFT_MASK) {
         jdouble t = dy;
         dy = dx;
         dx = t;
     }
-    if (jview) {
+
+    if (jview && (dx != 0 || dy != 0)) {
+        // Note: For gtk negatives values means UP/LEFT, JavaFx is opposite
+        dx = (dx != 0) ? dx * -1 : 0;
+        dy = (dy != 0) ? dy * -1 : 0;
```

Comment:
Is avoiding a possible negative 0 is really necessary? If not, this could simplify to `dx *= -1;` and `dy *= -1;`

Reply:
Will double check it, but I think I spaced out.

Reply:
dx = (dx != 0) ? -dx : 0;

also, is jdouble a floating point type?  if so, would that work correctly if dx/dy are very small?

Reply:
It's a floating point type.

Output is:
```
Delta: -0,000000, -0,058579
Delta: -0,000000, 0,029289

```